price-feeder/oracle/provider/gate.go

```diff
@@ -0,0 +1,534 @@
+	var candleMessage GateCandleResponse
+	if err := json.Unmarshal(bz, &candleMessage); err != nil {
+		return err
+	}
+
+	if candleMessage.Method != "kline.update" {
+		return fmt.Errorf("message is not a kline update")
+	}
+
+	var gateCandle GateCandle
+	if err := gateCandle.UnmarshalParams(candleMessage.Params); err != nil {
+		return err
+	}
+
+	p.setCandlePair(gateCandle)
+	return nil
+}
+
+func (p *GateProvider) setTickerPair(ticker GateTicker) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	p.tickers[ticker.Symbol] = ticker
+}
+
+func (p *GateProvider) setCandlePair(candle GateCandle) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	// convert gate timestamp seconds -> milliseconds
+	candle.TimeStamp = candle.TimeStamp * int64(time.Second/time.Millisecond)
+	staleTime := PastUnixTime(providerCandlePeriod)
+	candleList := []GateCandle{}
+
+	candleList = append(candleList, candle)
+	for _, c := range p.candles[candle.Symbol] {
+		if staleTime < c.TimeStamp {
+			candleList = append(candleList, c)
+		}
+	}
+	p.candles[candle.Symbol] = candleList
+}
+
+// subscribeTickerPairs write the subscription msg to the provider.
+func (p *GateProvider) subscribeTickerPairs(msg GateTickerSubscriptionMsg) error {
+	return p.wsClient.WriteJSON(msg)
+}
+
+// subscribeCandlePair write the subscription msg to the provider.
+func (p *GateProvider) subscribeCandlePair(msg GateCandleSubscriptionMsg) error {
+	return p.wsClient.WriteJSON(msg)
+}
+
+// setSubscribedPairs sets N currency pairs to the map of subscribed pairs.
+func (p *GateProvider) setSubscribedPairs(cps ...types.CurrencyPair) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	for _, cp := range cps {
+		p.subscribedPairs[cp.String()] = cp
+	}
+}
+
+func (p *GateProvider) resetReconnectTimer() {
+	p.reconnectTimer.Reset(gatePingCheck)
+}
+
+// reconnect closes the last WS connection and creates a new one. If there’s a
+// network problem, the system will automatically disable the connection. The
+// connection will break automatically if the subscription is not established or
+// data has not been pushed for more than 30 seconds. To keep the connection stable:
+// 1. Set a timer of N seconds whenever a response message is received, where N is
+// less than 30.
+// 2. If the timer is triggered, which means that no new message is received within
+// N seconds, send the String 'ping'.
+// 3. Expect a 'pong' as a response. If the response message is not received within
+// N seconds, please raise an error or reconnect.
+func (p *GateProvider) reconnect() error {
+	p.wsClient.Close()
+
+	p.logger.Debug().Msg("reconnecting websocket")
+	wsConn, _, err := websocket.DefaultDialer.Dial(p.wsURL.String(), nil)
+	if err != nil {
+		return fmt.Errorf("error reconnecting to Gate websocket: %w", err)
+	}
+	wsConn.SetPongHandler(p.pongHandler)
+	p.wsClient = wsConn
+
+	cps := make([]types.CurrencyPair, len(p.subscribedPairs))
+	iterator := 0
+	for _, cp := range p.subscribedPairs {
+		cps[iterator] = cp
+		iterator++
+	}
```

Comment:
- You could do the same as we do in the [binance](https://github.com/umee-network/umee/blob/d75ab49759e8bcdefb088d54d8cc12823e31d954/price-feeder/oracle/provider/binance.go#L320-L321)
```suggestion
	p.mtx.RLock()
	defer p.mtx.RUnlock()
	cps := mapPairsToSlice(p.subscribedPairs)
```

- or just add the mutex Read lock and unlock to it